### PR TITLE
[feat] support alloc contiguous pages at given addr, support 64G memo…

### PR DESCRIPTION
**Pull request by hky1999**
…ry range

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -20,13 +20,17 @@ slab = ["dep:slab_allocator"]
 buddy = ["dep:buddy_system_allocator"]
 
 allocator_api = []
+64G = []
```
Comment (Azure-stars) [inline Cargo.toml:23]: Can we use a more appropriate name? And maybe we need to specify that the `64G` belongs to `bitmap allocator`(The max pool size in `tlsf` is 8G)

File: src/bitmap.rs [hunk 2]
```diff
@@ -74,9 +79,15 @@ impl<const PAGE_SIZE: usize> PageAllocator for BitmapPageAllocator<PAGE_SIZE> {
     }
 
     fn dealloc_pages(&mut self, pos: usize, num_pages: usize) {
-        // TODO: not decrease `used_pages` if deallocation failed
-        self.used_pages -= num_pages;
-        self.inner.dealloc((pos - self.base) / PAGE_SIZE)
+        if match num_pages.cmp(&1) {
+            core::cmp::Ordering::Equal => self.inner.dealloc((pos - self.base) / PAGE_SIZE),
```
Comment (Azure-stars) [inline src/bitmap.rs:83]: can we replace it with `dealloc_contiguous((pos - self.base) / PAGE_SIZE, 1)`? 

then we can unify the `equal` and `greater` cases

Comment (hky1999) [reply src/bitmap.rs:83]: I'm just as curious as you about why it was written like this before.

File: Cargo.toml [hunk 3]
```diff
@@ -10,8 +10,8 @@ repository = "https://github.com/arceos-org/allocator"
 documentation = "https://arceos-org.github.io/allocator"
 
 [features]
-default = []
-full = ["bitmap", "tlsf", "slab", "buddy", "allocator_api"]
+default = ["page-alloc-4g"]
+full = ["bitmap", "tlsf", "slab", "buddy", "allocator_api","page-alloc-4g"]
```
Comment (equation314) [inline Cargo.toml:14]: Space before `"page-alloc-4g"`

File: Cargo.toml [hunk 4]
```diff
@@ -10,8 +10,8 @@ repository = "https://github.com/arceos-org/allocator"
 documentation = "https://arceos-org.github.io/allocator"
 
 [features]
-default = []
-full = ["bitmap", "tlsf", "slab", "buddy", "allocator_api"]
+default = ["page-alloc-4g"]
```
Comment (equation314) [inline Cargo.toml:13]: `page-alloc-256m`

File: src/bitmap.rs [hunk 5]
```diff
@@ -1,13 +1,24 @@
...
-//! TODO: adaptive size
 
 use bitmap_allocator::BitAlloc;
 
 use crate::{AllocError, AllocResult, BaseAllocator, PageAllocator};
 
-// Support max 1M * 4096 = 4GB memory.
-type BitAllocUsed = bitmap_allocator::BitAlloc1M;
+cfg_if::cfg_if! {
+if #[cfg(feature = "page-alloc-1t")] {
```
Comment (equation314) [inline src/bitmap.rs:8]: Indent

File: src/bitmap.rs [hunk 6]
```diff
@@ -65,18 +76,46 @@ impl<const PAGE_SIZE: usize> PageAllocator for BitmapPageAllocator<PAGE_SIZE> {
...
+        if align_pow2 % PAGE_SIZE != 0 {
+            return Err(AllocError::InvalidParam);
+        }
+        let align_pow2 = align_pow2 / PAGE_SIZE;
+        if !align_pow2.is_power_of_two() {
+            return Err(AllocError::InvalidParam);
+        }
+        let align_log2 = align_pow2.trailing_zeros() as usize;
+
+        let idx = (base - self.base) / PAGE_SIZE;
```
Comment (equation314) [inline src/bitmap.rs:103]: Check alignment of `base`

File: src/bitmap.rs [hunk 7]
```diff
@@ -65,18 +76,46 @@ impl<const PAGE_SIZE: usize> PageAllocator for BitmapPageAllocator<PAGE_SIZE> {
...
+            .map(|idx| idx * PAGE_SIZE + self.base)
+            .ok_or(AllocError::NoMemory)
+            .inspect(|_| self.used_pages += num_pages)
+    }
+
     fn dealloc_pages(&mut self, pos: usize, num_pages: usize) {
-        // TODO: not decrease `used_pages` if deallocation failed
-        self.used_pages -= num_pages;
-        self.inner.dealloc((pos - self.base) / PAGE_SIZE)
+        if self
```
Comment (equation314) [inline src/bitmap.rs:113]: Check alignment of `pos`

File: Cargo.toml (hunk 3 above)
Comment (hky1999) [reply Cargo.toml]: > Space before `"page-alloc-4g"`

fixed

File: src/bitmap.rs [hunk 8]
```diff
@@ -130,3 +137,13 @@ impl<const PAGE_SIZE: usize> PageAllocator for BitmapPageAllocator<PAGE_SIZE> {
         self.total_pages - self.used_pages
     }
 }
+
+impl<const PAGE_SIZE: usize> BitmapPageAllocator<PAGE_SIZE> {
+    /// Checks whether the address has the demanded alignment.
+    ///
+    /// Equivalent to `addr % align == 0`, but the alignment must be a power of two.
+    #[inline]
+    const fn is_aligned(base_addr: usize) -> bool {
```
Comment (equation314) [inline src/bitmap.rs:146]: No need to put it inside a impl block

Comment (hky1999) [reply src/bitmap.rs:146]: may we need the const type  `PAGE_SIZE`  associated with the `PageAllocator` ?

File: src/bitmap.rs (hunk 2 above)
Comment (equation314) [reply src/bitmap.rs]: `alloc_contigous` has a longer code path than `alloc` and may perform worse.

File: src/bitmap.rs [hunk 9]
```diff
@@ -37,11 +53,20 @@ impl<const PAGE_SIZE: usize> BitmapPageAllocator<PAGE_SIZE> {
 impl<const PAGE_SIZE: usize> BaseAllocator for BitmapPageAllocator<PAGE_SIZE> {
     fn init(&mut self, start: usize, size: usize) {
         assert!(PAGE_SIZE.is_power_of_two());
-        let end = super::align_down(start + size, PAGE_SIZE);
-        let start = super::align_up(start, PAGE_SIZE);
-        self.base = start;
+
+        // Range for real:  [align_up(start, PAGE_SIZE), align_down(start + size, PAGE_SIZE))
+        let start = crate::align_up(start, PAGE_SIZE);
+        let end = crate::align_down(start + size, PAGE_SIZE);
```
Comment (equation314) [inline src/bitmap.rs:59]: Swap these two lines, `start` should be the one in the argument.

File: src/bitmap.rs [hunk 10]
```diff
@@ -53,6 +78,10 @@ impl<const PAGE_SIZE: usize> PageAllocator for BitmapPageAllocator<PAGE_SIZE> {
     const PAGE_SIZE: usize = PAGE_SIZE;
 
     fn alloc_pages(&mut self, num_pages: usize, align_pow2: usize) -> AllocResult<usize> {
+        assert!(
```
Comment (equation314) [inline src/bitmap.rs:81]: Returns `Err(AllocError::InvalidParam)`

File: src/bitmap.rs (hunk 9 above)
Comment (hky1999) [reply src/bitmap.rs:59]: sorry, it's a stupid mistake

File: src/bitmap.rs [hunk 11]
```diff
@@ -65,18 +93,60 @@ impl<const PAGE_SIZE: usize> PageAllocator for BitmapPageAllocator<PAGE_SIZE> {
...
+    ) -> AllocResult<usize> {
+        if align_pow2 > MAX_ALIGN_1GB {
+            return Err(AllocError::InvalidParam);
+        }
+        if align_pow2 % PAGE_SIZE != 0 {
+            return Err(AllocError::InvalidParam);
+        }
+
+        // Check if the base address is aligned to the given alignment and the given PAGE_SIZE.
+        if !crate::is_aligned(base, align_pow2) || !crate::is_aligned(base, Self::PAGE_SIZE) {
```
Comment (equation314) [inline src/bitmap.rs:119]: Since `align_pow2 % PAGE_SIZE == 0`, `is_aligned(base, align_pow2)` implies `is_aligned(base, PAGE_SIZE)`.

**Review by equation314 — APPROVED** [review]

File: src/bitmap.rs (hunk 2 above)
Comment (hky1999) [reply src/bitmap.rs]: > `alloc_contiguous`/`dealloc_contiguous` has a longer code path than `alloc`/`dealloc` and may perform worse.

should we care for that, change it back to `dealloc`?

Comment (equation314) [reply src/bitmap.rs]: You decide

Comment (hky1999) [reply src/bitmap.rs:139]: > You decide

fixed

**Review by equation314 — APPROVED** [review]

**Pull request merged**